nexus/src/db/datastore.rs

```diff
@@ -707,17 +707,19 @@ impl DataStore {
             })?;
 
         // Update those datasets to which the regions belonged.
-        diesel::update(dataset_dsl::dataset)
-            .filter(dataset_dsl::id.eq(dataset_id))
-            .set(dataset_dsl::size_used.eq(dataset_dsl::size_used - size))
-            .execute_async(self.pool())
-            .await
-            .map_err(|e| {
-                Error::internal_error(&format!(
-                    "error updating dataset space: {:?}",
-                    e
-                ))
-            })?;
+        for (dataset_id, size) in datasets_id_and_size {
+            diesel::update(dataset_dsl::dataset)
+                .filter(dataset_dsl::id.eq(dataset_id))
+                .set(dataset_dsl::size_used.eq(dataset_dsl::size_used - size))
+                .execute_async(self.pool())
+                .await
+                .map_err(|e| {
+                    Error::internal_error(&format!(
+                        "error updating dataset space: {:?}",
+                        e
+                    ))
+                })?;
+        }
```

Comment:
I don't think this sequence of operations is idempotent, which I think makes it problematic in the context of a saga.

If a saga fails partway through operation, it may be re-tried. This can happen in the "action" direction (during the "disk delete" saga) or in the "undo" direction (during the "disk create" saga). Admittedly, it looks like this was true before this PR too, but I think it became more obvious to me with the addition of a `for` loop here.

The easiest short-term fix here is probably to move these operations into a transaction - otherwise:

- If we crash between "delete" and "update", the size accounting will be wrong
- If we crash between updating *some* but not *all* of the datasets, the accounting will be wrong

Reply:
Instead of adding and subtracting, `size_used` is now summed up after regions are inserted or deleted, meaning both region_allocate and regions_hard_delete should be idempotent.